File: Python/longest_palindrome.py
# https://leetcode.com/problems/longest-palindrome/description/
from utils.examination import Examiner, Testcase


class Solution:
    def longestPalindrome(self, s: str) -> int:
        if '' == s:
            return 0
        d: dict[str, int] = {}
        l: int = 0
        for c in s:
            if c not in d:
                d[c] = 1
                continue
            d[c] += 1
            if 2 == d[c]:
                l += 2
                d[c] = 0
        for k, v in d.items():
            if 1 <= v:
                l += 1
                break
        return l

def examine()->None:
    Examiner(Solution(), 'longestPalindrome').run((
        Testcase(('',), 0),
        Testcase(('a',), 1),
        Testcase(('abccccdd',), 7),
    ))

File: TypeScript/solutions/longest-palindrome.ts
// https://leetcode.com/problems/longest-palindrome/description/
import { Examiner, Testcase } from '../utils';


function longestPalindrome(s: string): number {
    if (s === '') return 0;
    const d = new Map<string, number>();
    let l = 0;
    for (const c of s) {
        const count = (d.get(c) ?? 0) + 1;
        d.set(c, count);
        if (count === 2) { l += 2; d.set(c, 0); }
    }
    for (const v of d.values()) { if (v >= 1) { l += 1; break; } }
    return l;
}

export function examine(): void {
    new Examiner(longestPalindrome).run([
        new Testcase([''], 0),
        new Testcase(['a'], 1),
        new Testcase(['abccccdd'], 7),
    ]);
}
